fem/bilininteg_ext.cpp

```diff
@@ -0,0 +1,1588 @@
+   {
+      double s_nodes[3*D1D*D1D*D1D];
+      for (int q = 0; q < NQ; ++q)
+      {
+         for (int d = q; d < ND; d += NQ)
+         {
+            s_nodes[0+d*3] = X(0,d,e);
+            s_nodes[1+d*3] = X(1,d,e);
+            s_nodes[2+d*3] = X(2,d,e);
+         }
+      }
+      for (int q = 0; q < NQ; ++q)
+      {
+         double J11 = 0; double J12 = 0; double J13 = 0;
+         double J21 = 0; double J22 = 0; double J23 = 0;
+         double J31 = 0; double J32 = 0; double J33 = 0;
+         for (int d = 0; d < ND; ++d)
+         {
+            const double b = B(q,d);
+            const double wx = G(0,q,d);
+            const double wy = G(1,q,d);
+            const double wz = G(2,q,d);
+            const double x = s_nodes[0+d*3];
+            const double y = s_nodes[1+d*3];
+            const double z = s_nodes[2+d*3];
+            J11 += (wx * x); J12 += (wx * y); J13 += (wx * z);
+            J21 += (wy * x); J22 += (wy * y); J23 += (wy * z);
+            J31 += (wz * x); J32 += (wz * y); J33 += (wz * z);
+            Xq(0,q,e) = b*x; Xq(1,q,e) = b*y; Xq(2,q,e) = b*z;
+         }
+         const double r_detJ = ((J11 * J22 * J33) + (J12 * J23 * J31) +
+                                (J13 * J21 * J32) - (J13 * J22 * J31) -
+                                (J12 * J21 * J33) - (J11 * J23 * J32));
+         J(0,0,q,e) = J11;
+         J(1,0,q,e) = J12;
+         J(2,0,q,e) = J13;
+         J(0,1,q,e) = J21;
+         J(1,1,q,e) = J22;
+         J(2,1,q,e) = J23;
+         J(0,2,q,e) = J31;
+         J(1,2,q,e) = J32;
+         J(2,2,q,e) = J33;
+         const double r_idetJ = 1.0 / r_detJ;
+         invJ(0,0,q,e) = r_idetJ * ((J22 * J33)-(J23 * J32));
+         invJ(1,0,q,e) = r_idetJ * ((J32 * J13)-(J33 * J12));
+         invJ(2,0,q,e) = r_idetJ * ((J12 * J23)-(J13 * J22));
+         invJ(0,1,q,e) = r_idetJ * ((J23 * J31)-(J21 * J33));
+         invJ(1,1,q,e) = r_idetJ * ((J33 * J11)-(J31 * J13));
+         invJ(2,1,q,e) = r_idetJ * ((J13 * J21)-(J11 * J23));
+         invJ(0,2,q,e) = r_idetJ * ((J21 * J32)-(J22 * J31));
+         invJ(1,2,q,e) = r_idetJ * ((J31 * J12)-(J32 * J11));
+         invJ(2,2,q,e) = r_idetJ * ((J11 * J22)-(J12 * J21));
+         detJ(q,e) = r_detJ;
+      }
+   });
+}
+
+typedef void (*fIniGeom)(const int ne,
+                         const double *B, const double *G, const double *X,
+                         double *x, double *J, double *invJ, double *detJ);
+
+static void PAGeom(const int dim,
+                   const int D1D,
+                   const int Q1D,
+                   const int NE,
+                   const double* B,
+                   const double* G,
+                   const double* X,
+                   double* Xq,
+                   double* J,
+                   double* invJ,
+                   double* detJ)
+{
+   const int id = (dim<<8)|(D1D)<<4|(Q1D);
+   static std::unordered_map<int, fIniGeom> call =
+   {
+      // 2D
+      {0x222,&PAGeom2D<2,2>},
```

Comment:
I think we want by default to have linear mesh to work for order 1 to 8 I think, so `0x222,&PAGeom2D<2,2>` to `0x229,&PAGeom2D<2,9>`, same for 3D.